Extra value types for xframe

linkRace opened the pull request:
Adds options to have xframe header set on blocklist/allowlist format or by custom defined functions.

Tests added for all cases.

linkRace commented:
@grawk looking to publish a `1.7.0` with this and #143 , let me know what you think.

maxmil7 commented on README.md:
```diff
@@ -62,8 +62,10 @@ __Please note that you must use [express-session](https://github.com/expressjs/s
 * `angular` Boolean - Optional. Shorthand setting to set `lusca` up to use the default settings for CSRF validation according to the [AngularJS docs]. Can be used with `cookie.options`.
 * `blocklist` Array or String - Optional. Allows defining a set of routes that will not have csrf protection.  All others will.
 * `allowlist` Array or String - Optional. Allows defining a set of routes that will have csrf protection.  All others will not.
+* `csrfFunction` Function - Returns `true` if route should require a token or `false` if route should not require a token.
```
we can mark it as Optional

linkRace replied:
will change

maxmil7 commented on lib/csrf.js:
```diff
@@ -92,17 +92,16 @@ module.exports = function (options) {
-                shouldBypass = req.path.indexOf(exclusion) === 0;
-                return shouldBypass;
-            });
+            shouldBypass = (blocklist.indexOf(req.path) !== -1);
```
this seems to change the functionality. For eg. earlier `/foo` in blocklist would return true for `/foo/` but this won't

linkRace replied:
ah, interesting. problem we were running into is that `/` in the blocklist would put all routes into the blocklist.

maxmil7 replied:
I think we can add an explicit check for `/` in config. For eg:
```
blocklist.some(function (exclusion) {
  shouldBypass = exclusion.length === 1 ? req.path === exclusion : req.path.indexOf(exclusion) === 0 :
  return shouldBypass;
})
     
```

maxmil7 replied:
@grawk any thoughts?

linkRace replied:
how about a case where the blocklist defines `/blocklist` and the route `/blocklist/extra/` is passed? 

should `/blocklist/extra/` be on the blocklist or only if `/blocklist/*` is specified?

maxmil7 commented on lib/csrf.js:
```diff
@@ -92,17 +92,16 @@ module.exports = function (options) {
-                return shouldBypass;
-            });
+            shouldBypass = (allowlist.indexOf(req.path) === -1);
+        }
```
same as previous comment

maxmil7 commented on lib/xframes.js:
```diff
@@ -6,9 +6,54 @@
+            xframeFunction = options.xframeFunction;
+            if (typeof xframeFunction === 'function') {
+                shouldBypass = false;
+                headerValue = xframeFunction(req);
```
we should set `shouldBypass` to true, if function returns false

maxmil7 commented on lib/xframes.js:
```diff
@@ -6,9 +6,54 @@
+            }
+
+            if (blocklist) {
+                shouldBypass = (blocklist.indexOf(req.path) !== -1);
```
same issue as one in csrf blocklist check

maxmil7 commented on test/csrf.js:
```diff
@@ -45,11 +45,11 @@ describe('CSRF', function () {
         var app = mock({
             csrf: {
-                blocklist: ['/blocklist1', '/blocklist2']
+                blocklist: ['/blocklist1', '/blocklist2', '/']
```
are we using blocklist1 anywhere?

maxmil7 commented on test/csrf.js:
```diff
@@ -81,7 +81,7 @@ describe('CSRF', function () {
         var app = mock({
             csrf: {
-                allowlist: ['/allowlist1', '/allowlist2']
+                allowlist: ['/', '/allowlist1', '/allowlist2']
```
please add a test for trailing `/` as well

maxmil7 commented on test/csrf.js:
```diff
@@ -110,10 +114,67 @@ describe('CSRF', function () {
+    it('should only require token on post to whitelist 2', function (done) {
+        var app = mock({
+            csrf: {
+                whitelist: ['/api/buttons', '/api/data']
```
is `/api/data` being used anywhere?

maxmil7 commented on test/xframe.js:
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
+            .expect(200)
+            .end(function (err, res) {
+                var isHeaderPresent = res.header['x-frame-options'] !== undefined;
+                assert(isHeaderPresent);
```
any reason not to use the expect syntax similar to line 57?

linkRace replied:
good point, will change

maxmil7 commented on test/xframe.js:
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
+            });
+    });
+
+    it('header (sameorigin) not on blocklist', function (done) {
```
this use case seems to be covered in previous test case

maxmil7 replied:
might be good to add a scenario where `allowlist` or `blocklist` is a string

maxmil7 commented on test/xframe.js:
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
+    });
+
+    it('header function', function (done) {
+        var config = { xframe: {xframeFunction: function (req) {
```
can we add a test for when function returns `false`

linkRace commented:
Believe I've addressed all the comments outside of the trailing / issue.
I could add in logic checking for a route name and/or a trailing /. Wouldn't want to break any existing functionality with a minor version bump, but I'm not sure that's the intended behavior. Let me know your thoughts.

maxmil7 commented:
Change partially implemented in #145 
Closing this PR for the time being.
